ext/ddtrace_profiling_native_extension/collectors_cpu_and_wall_time.c

```diff
@@ -0,0 +1,145 @@
+static VALUE collectors_cpu_and_wall_time_class = Qnil;
+
+struct cpu_and_wall_time_collector_state {
+  // Note: Places in this line that usually need to be changed when this struct is changed are tagged with
+  // "Update this when modifying state struct"
+  sampling_buffer *sampling_buffer;
+  VALUE recorder_instance;
+};
+
+static void cpu_and_wall_time_collector_typed_data_mark(void *state_ptr);
+static void cpu_and_wall_time_collector_typed_data_free(void *state_ptr);
+static VALUE _native_new(VALUE klass);
+static VALUE _native_initialize(VALUE self, VALUE collector_instance, VALUE recorder_instance, VALUE max_frames);
+static VALUE _native_sample(VALUE self, VALUE collector_instance);
+static void sample(VALUE collector_instance);
+static VALUE _native_thread_list(VALUE self);
+
+void collectors_cpu_and_wall_time_init(VALUE profiling_module) {
+  VALUE collectors_module = rb_define_module_under(profiling_module, "Collectors");
+  collectors_cpu_and_wall_time_class = rb_define_class_under(collectors_module, "CpuAndWallTime", rb_cObject);
+
+  // Instances of the CpuAndWallTime class are "TypedData" objects.
+  // "TypedData" objects are special objects in the Ruby VM that can wrap C structs.
+  // In this case, it wraps the cpu_and_wall_time_collector_state.
+  //
+  // Because Ruby doesn't know how to initialize native-level structs, we MUST override the allocation function for objects
+  // of this class so that we can manage this part. Not overriding or disabling the allocation function is a common
+  // gotcha for "TypedData" objects that can very easily lead to VM crashes, see for instance
+  // https://bugs.ruby-lang.org/issues/18007 for a discussion around this.
+  rb_define_alloc_func(collectors_cpu_and_wall_time_class, _native_new);
+
+  rb_define_singleton_method(collectors_cpu_and_wall_time_class, "_native_initialize", _native_initialize, 3);
+  rb_define_singleton_method(collectors_cpu_and_wall_time_class, "_native_sample", _native_sample, 1);
+  rb_define_singleton_method(collectors_cpu_and_wall_time_class, "_native_thread_list", _native_thread_list, 0);
+}
+
+// This structure is used to define a Ruby object that stores a pointer to a struct cpu_and_wall_time_collector_state
+// See also https://github.com/ruby/ruby/blob/master/doc/extension.rdoc for how this works
+static const rb_data_type_t cpu_and_wall_time_collector_typed_data = {
+  .wrap_struct_name = "Datadog::Profiling::Collectors::CpuAndWallTime",
+  .function = {
+    .dmark = cpu_and_wall_time_collector_typed_data_mark,
+    .dfree = cpu_and_wall_time_collector_typed_data_free,
+    .dsize = NULL, // We don't track profile memory usage (although it'd be cool if we did!)
+    //.dcompact = NULL, // FIXME: Add support for compaction
+  },
+  .flags = RUBY_TYPED_FREE_IMMEDIATELY
+};
+
+static void cpu_and_wall_time_collector_typed_data_mark(void *state_ptr) {
+  struct cpu_and_wall_time_collector_state *state = (struct cpu_and_wall_time_collector_state *) state_ptr;
```

Comment:
Types, who needs them when you can just cast from `void *`. 😉  (I know this is because of the callback signature requirements)

Reply:
People say C has is statically typed -- for me `void *` is the dynamic typing of C 🤣